bin/validator_tests.py

```diff
@@ -96,6 +98,70 @@ def windows_newline(x: str) -> Optional[str]:
     return generators
 
 
-GENERATORS: Final[
+INVALID_GENERATORS: Final[
     Sequence[tuple[str, str | Callable[[str], Optional[str]], Sequence[type[AnyValidator]]]]
-] = _list_generators()
+] = _list_invalid_generators()
+
+
+def _list_valid_generators() -> list[tuple[str, str | Callable[[str], Optional[str]], bool, bool]]:
+    generator_names: set[str] = set()
+    generators: list[tuple[str, str | Callable[[str], Optional[str]], bool, bool]] = []
+
+    T = TypeVar("T", bound=str | Callable[[str], Optional[str]])
+
+    # returns a function that can be called to register a new generator for valid tests
+    # can be used on its own or as decorator for a function
+    def register(
+        name: Optional[str] = None, space_change: bool = False, case_change: bool = False
+    ) -> Callable[[T], T]:
+        def decorator(func: T) -> T:
+            nonlocal name
+            if not isinstance(func, str) and not name:
+                assert hasattr(func, "__name__")
+                name = func.__name__
+            assert name
+            generator_names.add(name)
+            generators.append((name, func, space_change, case_change))
+            return func
+
+        return decorator
+
+    # simple generators
+    register("leading_space", space_change=True)(lambda x: f" {x}")
+    register("trailing_newline", space_change=True)(lambda x: f"{x}\n")
+
+    @register(space_change=True)
+    def all_newline(x: str) -> Optional[str]:
+        if " " not in x:
+            return None
+        return x.replace(" ", "\n")
+
+    @register(space_change=True)
+    def all_space(x: str) -> Optional[str]:
+        if "\n" not in x:
+            return None
+        return x.replace("\n", " ")
+
+    @register(space_change=True)
+    def append_space(x: str) -> Optional[str]:
+        if end_newline(x):
+            return None
+        return f"{x[:-1]} \n"
+
+    @register(space_change=True)
+    def windows_newline(x: str) -> Optional[str]:
+        if "\n" not in x or "\r" in x:
+            return None
+        return x.replace("\n", "\r\n")
+
+    @register(case_change=True)
+    def swap_case(x: str) -> Optional[str]:
+        y = x.swapcase()
+        return None if x == y else y
+
+    return generators
+
+
+VALID_GENERATORS: Final[Sequence[tuple[str, str | Callable[[str], Optional[str]], bool, bool]]] = (
```

Comment:
The type annotation does not show what the booleans are used for – you only know this at the moment they're constructed or deconstructed, because then you get the context from the variable names. Do we want to make this a named tuple or data class or something, to make it clear in the type definition that these elements are `name`, `generator`, `space_change`, and `case_change`? :slightly_smiling_face: 

(we could also do this for `INVALID_GENERATORS`, of course :smile:)

Reply:
All right, agreed! Then I'll merge so I can rebase my `validator_{flags,args}` rework on top of this :smile: Thanks again! :heart: